File: 3666. minimum-operations-to-equalize-binary-string.py
class Solution:
    def minOperations(self, s: str, k: int) -> int:
        n = len(s)
        ts = [SortedSet() for _ in range(2)]
        for i in range(n + 1):
            ts[i % 2].add(i)
        cnt0 = s.count('0')
        ts[cnt0 % 2].remove(cnt0)
        q = deque([cnt0])
        ans = 0
        while q:
            for _ in range(len(q)):
                cur = q.popleft()
                if cur == 0:
                    return ans
                l = cur + k - 2 * min(cur, k)
                r = cur + k - 2 * max(k - n + cur, 0)
                t = ts[l % 2]
                j = t.bisect_left(l)
                while j < len(t) and t[j] <= r:
                    q.append(t[j])
                    t.remove(t[j])
            ans += 1
        return -1

if __name__ == "__main__":
    from collections import deque
    from sortedcontainers import SortedSet
    
    sol = Solution()
    
    # Example 1:
    s = "110"
    k = 1
    output = sol.minOperations(s, k)
    print(f"Example 1: Input: s = {s}, k = {k}")
    print(f"Output: {output}")
    print(f"Expected: 1")
    print()
    
    # Example 2:
    s = "0101"
    k = 3
    output = sol.minOperations(s, k)
    print(f"Example 2: Input: s = {s}, k = {k}")
    print(f"Output: {output}")
    print(f"Expected: 2")
    print()
    
    # Example 3:
    s = "101"
    k = 2
    output = sol.minOperations(s, k)
    print(f"Example 3: Input: s = {s}, k = {k}")
    print(f"Output: {output}")
    print(f"Expected: -1")

File: 3666. minimum-operations-to-equalize-binary-string.java
class Solution {
    public int minOperations(String s, int k) {
        int n = s.length();

        TreeSet<Integer>[] ts = new TreeSet[2];
        Arrays.setAll(ts, i -> new TreeSet<>());

        for (int i = 0; i <= n; i++) {
            ts[i % 2].add(i);
        }

        int cnt0 = 0;
        for (char c : s.toCharArray()) {
            if (c == '0') {
                cnt0++;
            }
        }

        ts[cnt0 % 2].remove(cnt0);

        Deque<Integer> q = new ArrayDeque<>();
        q.offer(cnt0);

        int ans = 0;
        while (!q.isEmpty()) {
            for (int size = q.size(); size > 0; --size) {
                int cur = q.poll();
                if (cur == 0) {
                    return ans;
                }

                int l = cur + k - 2 * Math.min(cur, k);
                int r = cur + k - 2 * Math.max(k - n + cur, 0);

                TreeSet<Integer> t = ts[l % 2];

                Integer next = t.ceiling(l);
                while (next != null && next <= r) {
                    q.offer(next);
                    t.remove(next);
                    next = t.ceiling(l);
                }
            }
            ans++;
        }

        return -1;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1:
        String s1 = "110";
        int k1 = 1;
        int output1 = sol.minOperations(s1, k1);
        System.out.println("Example 1: Input: s = " + s1 + ", k = " + k1);
        System.out.println("Output: " + output1);
        System.out.println("Expected: 1");
        System.out.println();
        
        // Example 2:
        String s2 = "0101";
        int k2 = 3;
        int output2 = sol.minOperations(s2, k2);
        System.out.println("Example 2: Input: s = " + s2 + ", k = " + k2);
        System.out.println("Output: " + output2);
        System.out.println("Expected: 2");
        System.out.println();
        
        // Example 3:
        String s3 = "101";
        int k3 = 2;
        int output3 = sol.minOperations(s3, k3);
        System.out.println("Example 3: Input: s = " + s3 + ", k = " + k3);
        System.out.println("Output: " + output3);
        System.out.println("Expected: -1");
    }
}